api/metric/sdkhelpers.go

```diff
@@ -95,3 +132,146 @@ func Configure(opts []Option) Config {
+// otherwise the error passes through.
+func WrapInt64CounterInstrument(syncInst SynchronousImpl, err error) (Int64Counter, error) {
+	common, err := checkSynchronous(syncInst, err)
+	return Int64Counter{synchronousInstrument: common}, err
+}
+
+func (m *wrappedMeterImpl) NewFloat64Counter(name string, opts ...Option) (Float64Counter, error) {
+	return WrapFloat64CounterInstrument(
+		m.newSynchronous(name, CounterKind, core.Float64NumberKind, opts))
+}
+
+// WrapInt64CounterInstrument returns an `Float64Counter` from from a
+// `SynchronousImpl`.  An error will be generated if the
+// `SynchronousImpl` is nil (in which case a No-op is substituted),
+// otherwise the error passes through.
+func WrapFloat64CounterInstrument(syncInst SynchronousImpl, err error) (Float64Counter, error) {
+	common, err := checkSynchronous(syncInst, err)
+	return Float64Counter{synchronousInstrument: common}, err
+}
+
+func (m *wrappedMeterImpl) NewInt64Measure(name string, opts ...Option) (Int64Measure, error) {
+	return WrapInt64MeasureInstrument(
+		m.newSynchronous(name, MeasureKind, core.Int64NumberKind, opts))
+}
+
+// WrapInt64MeasureInstrument returns an `Int64Measure` from from a
+// `SynchronousImpl`.  An error will be generated if the
+// `SynchronousImpl` is nil (in which case a No-op is substituted),
+// otherwise the error passes through.
+func WrapInt64MeasureInstrument(syncInst SynchronousImpl, err error) (Int64Measure, error) {
+	common, err := checkSynchronous(syncInst, err)
+	return Int64Measure{synchronousInstrument: common}, err
+}
+
+func (m *wrappedMeterImpl) NewFloat64Measure(name string, opts ...Option) (Float64Measure, error) {
+	return WrapFloat64MeasureInstrument(
+		m.newSynchronous(name, MeasureKind, core.Float64NumberKind, opts))
+}
+
+// WrapFloat64MeasureInstrument returns an `Float64Measure` from from a
+// `SynchronousImpl`.  An error will be generated if the
+// `SynchronousImpl` is nil (in which case a No-op is substituted),
+// otherwise the error passes through.
+func WrapFloat64MeasureInstrument(syncInst SynchronousImpl, err error) (Float64Measure, error) {
+	common, err := checkSynchronous(syncInst, err)
+	return Float64Measure{synchronousInstrument: common}, err
+}
+
+func (m *wrappedMeterImpl) newAsynchronous(name string, mkind Kind, nkind core.NumberKind, opts []Option, callback func(func(core.Number, LabelSet))) (AsynchronousImpl, error) {
+	return m.impl.NewAsynchronousInstrument(
+		NewDescriptor(name, mkind, nkind, opts...),
+		callback)
+}
+
+func (m *wrappedMeterImpl) RegisterInt64Observer(name string, callback Int64ObserverCallback, opts ...Option) (Int64Observer, error) {
+	if callback == nil {
+		return NoopMeter{}.RegisterInt64Observer("", nil)
+	}
+	result := &int64ObserverResult{}
+	return WrapInt64ObserverInstrument(
+		m.newAsynchronous(name, ObserverKind, core.Int64NumberKind, opts,
+			func(observe func(core.Number, LabelSet)) {
+				result.observe = observe
```

Comment:
Could the result be constructed inside the callback? If it implements the result interface by value, then there should be no extra allocation, right?